doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,307 @@
+
+    % PARAM-NAME type [flags]
+    code to process param
+
+    %
+
+    epilogue
+```
+
+- the % characters at start of line
+
+- FUNCTION-BASE-NAME is the name the functions will derive from e.g. "kdf_scrypt" to this will be appended _[gs]et[_ctx]_params
+
+- type is the standard OSSL_PARAM types with additions for basic C types
+  - exact list not certain
+
+- Function-flags include:
+  - nostatic: do not make the function static
+
+- Flags include:
+  - nocode: do not generate code for this parameter
+    - This allows, e.g., two different types for a parameter (int & string)
+  - unpublished: do not generate this parameter in the gettable/settable list
+    - "engine" is the only one like this
+
+
+The idea is that the gettable and get functions will be simultaneously generated
+along with a perfect hash to decode the legal parameters quickly.
+
+
+
+A worked example for scrypt:
+```
+{- generate_ctx_set(q{
+%% kdf_scrypt
+    /* prologue */
+    KDF_SCRYPT *ctx = vctx;
+    uint64_t u64_value;
+
+% OSSL_KDF_PARAM_PASSWORD       octet_string
+    if (!scrypt_set_membuf(&ctx->pass, &ctx->pass_len, p))
+        return 0;
+
+% OSSL_KDF_PARAM_SALT           octet_string
+    if (!scrypt_set_membuf(&ctx->salt, &ctx->salt_len, p))
+        return 0;
+
+% OSSL_KDF_PARAM_SCRYPT_N       uint64_t
+    if (!OSSL_PARAM_get_uint64(p, &u64_value)
+        || u64_value <= 1
+        || !is_power_of_two(u64_value))
+        return 0;
+    ctx->N = u64_value;
+
+% OSSL_KDF_PARAM_SCRYPT_R       uint32_t
+    if (!OSSL_PARAM_get_uint64(p, &u64_value) || u64_value < 1)
+        return 0;
+    ctx->r = u64_value;
+
+% OSSL_KDF_PARAM_SCRYPT_P       uint32_t
+    if (!OSSL_PARAM_get_uint64(p, &u64_value) || u64_value < 1)
+        return 0;
+    ctx->p = u64_value;
+
+% OSSL_KDF_PARAM_SCRYPT_MAXMEM  uint64_t
+    if (!OSSL_PARAM_get_uint64(p, &u64_value) || u64_value < 1)
+        return 0;
+    ctx->maxmem_bytes = u64_value;
+
+% OSSL_KDF_PARAM_PROPERTIES     utf8_string
+    if (p->data_type != OSSL_PARAM_UTF8_STRING
+        || !set_property_query(ctx, p->data)
+        || !set_digest(ctx))
+        return 0;
+%
+    /* epilogue */
+    return 1;
+});
+
+generate_ctx_get(q{
+    /* empty prologue */
+% OSSL_KDF_PARAM_SIZE           size_t
+    return OSSL_PARAM_set_size_t(p, SIZE_MAX);
+%
+    /* epilogue */
+    return -2;
+});
+-}
+```
+
+Would generate something along the lines of:
+```c
+enum {
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_SALT = 0,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_PROPERTIES = 1,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_SCRYPT_P = 2,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_SCRYPT_MAXMEM = 3,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_SCRYPT_R = 4,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_PASSWORD = 5,
+    kdf_scrypt_ctx_set_param_OSSL_KDF_PARAM_SCRYPT_N = 6,
+    kdf_scrypt_ctx_set_MAX
+};
+
+static int kdf_scrypt_ctx_set_hash (const char *str, size_t len)
+{
+    static const unsigned char asso_values[] = {
```

Comment:
Would a range check suffice instead of a full 256 byte table?

Reply:
It would be slower but would certainly be possible.  gperf has an option to output 128 byte tables & we could reduce the character set further to alphanumerics, _ and -.

One thing to be cognisant of is that, in general, we will always be hashing strings that do match and that mismatches are very very much the exception.

Reply:
If I thought it was future safe, I'd omit the strcmp entirely.
Unfortunately it isn't.